src/main/scala/org/ergoplatform/Validation.scala

```diff
@@ -0,0 +1,153 @@
+  */
+case class  ReplacedRule(newRuleId: Short) extends RuleStatus
+
+/** The status of the rule whose parameters are changed via soft-fork extensions.
+  * @param newValue  new value of block extension value with key == rule.id
+  */
+case class  ChangedRule(newValue: Array[Byte]) extends RuleStatus
+
+/** Base class for different validation rules registered in ValidationRules.currentSettings.
+  * Each rule is identified by `id` and have a description.
+  * Validation logic is implemented by `apply` methods of derived classes.
+  */
+case class ValidationRule(
+  id: Short,
+  description: String
+) {
+  /** Can be used in derived classes to implemented validation logic. */
+  protected def validate[T](
+        vs: ValidationSettings, condition: => Boolean,
+        cause: => Throwable, args: Seq[Any], block: => T): T = {
+    val status = vs.getStatus(this.id)
+    status match {
+      case None =>
+        throw new InterpreterException(s"ValidationRule $this not found in validation settings")
+      case Some(status) =>
+        if (condition)
+          block
+        else {
+          status match {
+            case EnabledRule =>
+              throw new ValidationException("Validation failed.", this, args, Option(cause))
+            case r: ReplacedRule =>
+              throw ReplacedRuleException(vs, this, r)
+            case c: ChangedRule =>
+              throw ChangedRuleException(vs, this, c)
+          }
+        }
+    }
+  }
+}
+
+/** Base class for all exception which may be thrown by validation rules. */
+case class ValidationException(message: String, rule: ValidationRule, args: Seq[Any], cause: Option[Throwable] = None)
+  extends Exception(message, cause.orNull)
+
+/** Instances of this class are used as messages to communicate soft-fork information,
+  * from the context where the soft-fork condition is detected (such as in ValidationRules),
+  * up the stack to the point where it is clear how to handle it.
+  * Some messages of this kind are not handled, it which case a new Exception is thrown
+  * and this instance should be attached as `cause` parameter. */
+class SoftForkException(message: String) extends Exception(message) {
+  /** This override is required as an optimization to avoid spending time on recording stack trace.
+    * This makes throwing exceptions almost as fast as usual return of a method.
+    */
+  override def fillInStackTrace(): Throwable = this
+}
+
+case class ReplacedRuleException(vs: ValidationSettings, replacedRule: ValidationRule, replacement: ReplacedRule)
+  extends SoftForkException(s"Rule ${replacedRule.id} was replaced with ${replacement.newRuleId}")
+
+case class ChangedRuleException(vs: ValidationSettings, changedRule: ValidationRule, change: ChangedRule)
+  extends SoftForkException(s"Rule ${changedRule.id} was changed with ${change}")
+
+/**
+  * Configuration of validation.
+  */
+abstract class ValidationSettings {
```

Comment:
How should `ValidationRule` and `RuleStatus`it be specified on proposing a soft-fork?

Reply:
amended the class doc.